preload/main.c

```diff
@@ -22,28 +23,40 @@ int ioctl (int d, int request, char *argp) {
     if (orig_ioctl == NULL) {
         orig_ioctl = dlsym(RTLD_NEXT, "ioctl");
     }
+
+    /* We only care for TIOCGWINSZ ioctls */
+    if (request != 0x5413) {
+        return orig_ioctl(d, request, argp);
+    }
+
     static int max_rows = -1;
+
+    /* ioctl gets called once before the perl module had the time to determine
+     * the right size! Leave max_rows negative to indicat that it still needs to
+     * be read from the SHELLEX_SIZE_FILE */
+
     if (max_rows < 0 ) {
-        char *str = getenv("SHELLEX_MAX_ROWS");
+
+      char *fname = getenv("SHELLEX_SIZE_FILE");
+      FILE *stream = fopen (fname,"r");
+      char str[5] = "-500";
+      if (stream) {
+        char *ret = fgets(str,6,stream);
+        fclose(stream);
         if (str != NULL) {
-            max_rows = atoi(str);
+          /* this may be -500 */
+          max_rows = atoi(str);
         }
+      }
     }
 
-    // We only care for TIOCSWINSZ ioctls
-    if (request != 0x5414) {
-        return orig_ioctl(d, request, argp);
-    }
+    int retval = orig_ioctl(d, request, (char *)argp);
+    struct winsize *ws = (struct winsize *)argp;
```

Comment:
unless ioctl's are allowed to modify their arguments (I don't think they are, and even if…), it's safer to keep a copy on the stack, as we did before. argp could be in read-only memory, for all we know, in which case this will segfault. Making a stack-copy is free, so we should go for it.